### PR TITLE
Support configurable policy for query node, Add user level schedule policy

### PR DESCRIPTION
**Pull request by chyezh**
Related issue #23716

- Support configurable policy for query node.
- Add new user level schedule policy in query node.

### CODE REVIEW

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

**Comment by chyezh** [comment]
rerun ut

File: internal/util/paramtable/component_param.go [hunk 1]
```diff
@@ -1307,6 +1314,24 @@ func (p *queryNodeConfig) initDiskCapacity() {
 	p.DiskCapacityLimit = diskSize * 1024 * 1024 * 1024
 }
 
+func (p *queryNodeConfig) initScheduleReadPolicy() {
+	validPolicies := []string{
+		"default",
+		"fifo",
+		"user-task-polling",
+	}
+	p.ScheduleReadPolicy = p.Base.Get("queryNode.scheduler.scheduleReadPolicy")
```
Comment (czs007) [inline internal/util/paramtable/component_param.go:1323]: use loadWithDefault instead.

File: internal/util/paramtable/component_param.go [hunk 2]
```diff
@@ -1307,6 +1314,24 @@ func (p *queryNodeConfig) initDiskCapacity() {
 	p.DiskCapacityLimit = diskSize * 1024 * 1024 * 1024
 }
 
+func (p *queryNodeConfig) initScheduleReadPolicy() {
+	validPolicies := []string{
+		"default",
```
Comment (czs007) [inline internal/util/paramtable/component_param.go:1319]: unnecessary?

File: internal/querynode/scheduler_policy.go [hunk 3]
```diff
@@ -2,23 +2,305 @@ package querynode
...
+	mergeTask(rt readTask) bool
+
+	// Schedule new task to run.
+	schedule(targetCPUUsage int32, maxNum int32) ([]readTask, int32)
+}
+
+// Create a new schedule policy.
+func newReadScheduleTaskPolicy(policyName string) scheduleReadPolicy {
+	switch policyName {
+	case "":
```
Comment (czs007) [inline internal/querynode/scheduler_policy.go:34]: unnecessary

File: internal/querynode/scheduler_policy.go [hunk 4]
```diff
@@ -2,23 +2,305 @@ package querynode
...
+		fallthrough
+	case "fifo":
```
Comment (czs007) [inline internal/querynode/scheduler_policy.go:36]: It is recommended to define "fifo" and "user-task-polling" as Const variables under the querynode package.

File: internal/querynode/scheduler_policy.go [hunk 5]
```diff
@@ -2,23 +2,305 @@ package querynode
...
+		p.route[username] = newRing
+		if p.checkpoint == nil {
+			// Create new ring if not exist.
+			p.checkpoint = newRing
+		} else {
+			// Add the new ring before the checkpoint.
+			p.checkpoint.Prev().Link(newRing)
+		}
+	}
+	p.taskCount++
```
Comment (czs007) [inline internal/querynode/scheduler_policy.go:156]: Is this taskCount concurrency safe?

File: internal/querynode/scheduler_policy.go [hunk 6]
```diff
@@ -2,23 +2,305 @@ package querynode
...
+			}
+
+			next := checkpoint.Next()
+			// Find task in this queue.
+			taskQueue := checkpoint.Value.(*userBasedTaskQueue)
+
+			// empty task queue for this user.
+			if taskQueue.len() == 0 {
+				// expire the queue.
+				if taskQueue.expire(time.Minute) {
```
Comment (czs007) [inline internal/querynode/scheduler_policy.go:216]: make time.Minute configurage?

File: internal/querynode/scheduler_policy.go (hunk 5 above)
Comment (chyezh) [reply internal/querynode/scheduler_policy.go:156]: Schedule policy is always run in a single goroutine, no race condition here.

File: internal/util/paramtable/component_param.go [hunk 7]
```diff
@@ -1346,6 +1362,35 @@ func (p *queryNodeConfig) initDiskCapacity() {
 	p.DiskCapacityLimit = diskSize * 1024 * 1024 * 1024
 }
 
+func (p *queryNodeConfig) initScheduleReadPolicy() {
+	const (
+		policyNameFIFO            = "fifo"
+		policyNameUserTaskPolling = "user-task-polling"
+	)
+
+	validPolicies := []string{
```
Comment (czs007) [inline internal/util/paramtable/component_param.go:1371]: no need to check

File: internal/util/paramtable/component_param.go [hunk 8]
```diff
@@ -1346,6 +1362,35 @@ func (p *queryNodeConfig) initDiskCapacity() {
...
+	p.ScheduleReadPolicy.Name = p.Base.LoadWithDefault("queryNode.scheduler.scheduleReadPolicy.name", "fifo")
+
+	isValidName := false
+	for _, policy := range validPolicies {
+		if policy == p.ScheduleReadPolicy.Name {
+			isValidName = true
+			break
+		}
+	}
+	if !isValidName {
```
Comment (czs007) [inline internal/util/paramtable/component_param.go:1384]: no need to check

File: internal/util/paramtable/component_param.go [hunk 9]
```diff
@@ -1346,6 +1362,35 @@ func (p *queryNodeConfig) initDiskCapacity() {
...
+		panic(fmt.Sprintf("invalid scheduleReadPolicy of query node, valid policy:%v", validPolicies))
+	}
+
+	switch p.ScheduleReadPolicy.Name {
```
Comment (czs007) [inline internal/util/paramtable/component_param.go:1388]: no need to switch, just parse and load

File: internal/querynode/scheduler_policy.go (hunk 3 above)
Comment (chyezh) [reply internal/querynode/scheduler_policy.go:39]: for passing old unittest

**Pull request merged**